photon-lib/src/main/scala/com/linkedin/photon/ml/function/glm/HessianVectorAggregator.scala

```diff
@@ -15,159 +15,276 @@
+ * loss function (possibly with normalization) on a dataset and some vector d.
  *
- * Some logic of Hessian vector multiplication is the same for gradient aggregation, so this class inherits
- * ValueAndGradientAggregator.
+ * This class is implemented as a convenience to more efficiently compute this value, rather than computing the entire
+ * Hessian matrix and performing a multiplication operation with vector d afterwards.
  *
- * @param func A single loss function for the generalized linear model
- * @param dim The dimension (number of features) of the aggregator
+ * @param lossFunction The GLM loss function
+ * @param coefficients The coefficients for the GLM
+ * @param dVector The vector d to multiply with the Hessian matrix
+ * @param normalizationContext The feature normalization information
  */
-@SerialVersionUID(2L)
-protected[ml] class HessianVectorAggregator(func: PointwiseLossFunction, dim: Int)
-  extends ValueAndGradientAggregator(func, dim) {
-
-  // effectiveMultiplyVector_j = factor_j * multiplyVector
-  // This intermediate vector helps to facilitate calculating
-  //    \sum_k (x_{ki} - shift_k) * factor_k * multiplyVector_k
-  //  = \sum_k (x_{ki} - shift_k) * effectiveMultiplyVector_k
-  // This vector is data point independent.
-  @transient var effectiveMultiplyVector: Vector[Double] = _
-
-  // featureVectorProductShift = \sum_k shift_k * effectiveMultiplyVector_k
-  // This intermediate value helps to facilitate calculating
-  //     \sum_k (x_{ki} - shift_k) * factor_k * multiplyVector_k
-  //   = \sum_k x_{ki} * effectiveMultiplyVector_k - featureVectorProductShift
-  // This value is data point independent.
-  @transient var featureVectorProductShift: Double = _
+@SerialVersionUID(3L)
+protected[ml] class HessianVectorAggregator(
+    lossFunction: PointwiseLossFunction,
+    coefficients: BroadcastWrapper[Vector[Double]],
+    dVector: BroadcastWrapper[Vector[Double]],
+    normalizationContext: BroadcastWrapper[NormalizationContext])
+  extends ValueAndGradientAggregator(lossFunction, coefficients, normalizationContext) {
 
   /**
-   * Initialize the aggregator with proper multiply vector and product shifts if normalization is used.
+   * Expanding upon the documentation in the [[ValueAndGradientAggregator]], the equation for computing the Hessian
+   * matrix for a particular set of coefficients over a dataset is as follows:
    *
-   * @param coef The current model coefficients
-   * @param multiplyVector The Hessian multiplication vector
-   * @param normalizationContext The normalization context
+   * H_jk = ∂^2^ L(w, X, y, u, o) / (∂ w_j)(∂ w_k)
+   *      = \sum_i ∂^2^ (u_i * l(z_i, y_i)) / (∂ w_j)(∂ w_k)
+   *      = \sum_i (∂^2^ l(z_i, y_i) / (∂ z_i)^2^) * u_i * X_i_j * X_i_k
+   *
+   * Let v = H * d, where d is some vector. Then v_j = H_j * d, where:
+   *
+   * H_j = /sum_i H_i_j         =>
```

Comment:
Should the summation notation be unified as `\sum`?

Reply:
That's a typo - will fix